### DISCUSSION TITLE
Where is Write_ChckOutLst.m file? (and other questions)

### DISCUSSION THREAD

**Discussion by aitorplaza** [OpenFAST/openfast#879] [Q&A]
Hi to everybody

I'm want to create a Pull Request and I've seen that some code parts have been generated with _Write_ChckOutLst.m_ . Where is it? I've been searching on this repository and on _matlab-toolbox_ repository but I don't see it. 

If I want to create a PR, is there any guideline to be followed? I guess that it must be done to  _dev_ branch, right? 

Thanks in advance

**Comment by bjonkman** [comment]
https://github.com/OpenFAST/matlab-toolbox/blob/master/ProgrammingTools/Write_ChckOutLst.m

**Comment by aitorplaza** [comment]
> https://github.com/OpenFAST/matlab-toolbox/blob/master/ProgrammingTools/Write_ChckOutLst.m

Hi again

I need this output parameter, YawBrTV[x,y,z]p so I've added the needed code to ElastoDyn_IO.f90 and ElastoDyn.f90. 

Those variables are obtained as:

```
   m%AllOuts(YawBrTVxp) =  DOT_PRODUCT( m%RtHS%LinVelEO, m%CoordSys%b1 )
   m%AllOuts(YawBrTVyp) = -DOT_PRODUCT( m%RtHS%LinVelEO, m%CoordSys%b3 )
   m%AllOuts(YawBrTVzp) =  DOT_PRODUCT( m%RtHS%LinVelEO, m%CoordSys%b2 )  
```

and i think that this is correct. 

I've run _Write_ChckOutLst.m_ in order to obtain the lists of parameters (ParamUnitsAry,ParamIndxAry, ValidParamAry ) and on the output files (ElastoDyn_Parameters.f90 and ElastoDyn_SetOutParam.f90) I'm getting parameter per row wrong. For example on _ValidParamAry_ (ElasoDyn.f90) , 8 parameters for row and not 7, as the actual code has. 

I suppose that, following the same code style, you want 7 parameters per row, right?

On the other hand, my changes work properly  but when compiling i'm getting the next warning, any idea about this?

```
Scanning dependencies of target elastodynlib
[ 45%] Building Fortran object modules/elastodyn/CMakeFiles/elastodynlib.dir/src/ElastoDyn_IO.f90.o
[ 46%] Building Fortran object modules/elastodyn/CMakeFiles/elastodynlib.dir/src/ElastoDyn.f90.o
/home/aitor/openfast/modules/elastodyn/src/ElastoDyn.f90:2290:0:

          ErrMsg = 'Error in Init_DOFparameters: '//TRIM(ErrMsg)
 
Warning: ‘__builtin_memset’: specified size 18446744073709551587 exceeds maximum object size 9223372036854775807 [-Wstringop-overflow=]
[ 46%] Linking Fortran static library libelastodynlib.a
```

Thanks in advance and if you think that this code can be useful, I can do a PR.

**Comment by bjonkman** [comment]
I updated the `write_ChckOutLst.m` algorithm a while ago to put more parameters per row if they fit (depending on the max size of the parameter names), so I think it is good that it has 8 instead of 7 now.

The warning you see is common on some compilers. It's telling you that the string might not fit in `ErrMsg`, depending on how many characters are returned from the `trim()` function. In Fortran, this could result in missing the last part of an error message if the length of `ErrMsg` is not defined to be large enough. In other languages, this could cause some memory issues. I wouldn't worry about it here.

**Comment by jjonkman** [comment]
Dear @aitorplaza,

Regarding the new yaw-bearing velocity outputs you've added, I agree with your calculations.  And I agree that these would be useful to add to ElastoDyn through a PR.

Thanks!

**Comment by aitorplaza** [comment]
Thank you to both of you. You are doing an awesome work with OpenFAST, it is very helpful for some of us.

I will review my changes again and I will create a PR.

**Comment by andrew-platt** [comment]
@aitorplaza , please tag me in your PR when you make it.  I can help you with any details there.

**Comment by aitorplaza** [comment]
Dear developers ( @andrew-platt )

I've reopened this thread just to not create a new one and fill the issues forum with more info about the same thing.

I've been "playing" with my last PR and I've noticed that I need more output variables, such as `YawBrTV[x,y,x]t`, `YawBrTA[x,y,x]t` and the corresponding yaw absolute rotations, angular velocities and accelerations about `t` and `p` basis. I need to add them to my code so if you are interested I can create again one new PR (Sorry for performing this in two parts). I think that this will complete all the yaw kinematics.  

Tell me if you agree with a new Pull Request.

thanks in advance.

**Comment by andrew-platt** [comment]
Hi @aitorplaza.  We would certainly be interested in adding these outputs.  Since you are finding these useful, it is very likely that others would find them useful as well.

So go ahead and put together another PR and we will merge it in.  Also, no problem with multiple pull requests -- that often happens in code development and really doesn't complicate anything for us.

**Comment by jjonkman** [comment]
Hi @aitorplaza,

Do you mean "t" and "p" or did you mean "i" and "p"?  I'm not sure it makes sense to output the absolute velocities and accelerations in "t", which is the tower-base coordinate system that itself may move due to motion of the foundation or offshore substructure.  Outputting the absolute velocities and accelerations in "i"--the inertial frame coordinate system--would make sense to me.

Best regards,

**Comment by aitorplaza** [comment]
Hi @jjonkman 

Thank you for your response and, yes, you are right, i was thinking about onshore machines, where `t` and `i` should be same, right?. So yes, I need the absolute velocities and accelerations expressed in the inertial frame. I've messed up because OpenFAST also computes `YawBrTD[x,y,z]t`.

But don't worry, I've playing with MATLAB and now I can compute "a posteriori" the needed velocities (`YawBrTV[x,y,x]t` in my case also `YawBrTV[x,y,x]i`). I've done it by two methods:

- by numerical differentiation of `YawBrTD[x,y,x]t` vector, it should be `YawBrTD[x,y,x]i` for offshore machines
- by change of basis of `YawBrTV[x,y,x]p` vector
 
Both methods give me the same result velocity vector, so I think that they are correct. 


If you think that  `YawBrTV[x,y,x]i`,  `YawBrTA[x,y,x]i` and the related yaw rotations can be useful, I will be glad to collaborate and can create a PR.

**Comment by jjonkman** [comment]
Hi @aitorplaza,

For an onshore wind turbine with a fixed foundation, I agree that "t" and "i" are the same.  But to be applicable for other support structures, I would recommend adding the outputs you want in "i" instead of "t".   It would be great if you would add these outputs and share through a PR.

Best regards,

**Comment by aitorplaza** [comment]
Hi @jjonkman 

I'll give a look to it and try to implement those kinematic entities. If I have any doubt, is this forum the right place to ask of better the old NREL forum?

**Comment by jjonkman** [comment]
Hi @aitorplaza,

 This is the right place for OpenFAST development-related Q&A (source code, compiling, etc.).  The old forum (https://wind.nrel.gov/forum/wind/) is great for general OpenFAST usage-related Q&A.

Best regards,
